crates/test-programs/src/bin/http_0_3_outbound_request_content_length.rs

```diff
@@ -0,0 +1,108 @@
+use futures::SinkExt as _;
+use test_programs::p3::wasi::http::types::{ErrorCode, Headers, Method, Request, Scheme, Trailers};
+use test_programs::p3::{wit_future, wit_stream};
+use wit_bindgen_rt::async_support::{FutureWriter, StreamWriter};
+
+struct Component;
+
+test_programs::p3::export!(Component);
+
+fn make_request() -> (
+    Request,
+    StreamWriter<u8>,
+    FutureWriter<Result<Option<Trailers>, ErrorCode>>,
+) {
+    let (contents_tx, contents_rx) = wit_stream::new();
+    let (trailers_tx, trailers_rx) = wit_future::new();
+    let (request, _) = Request::new(
+        Headers::from_list(&[("Content-Length".to_string(), b"11".to_vec())]).unwrap(),
+        Some(contents_rx),
+        trailers_rx,
+        None,
+    );
+
+    request.set_method(&Method::Post).expect("setting method");
+    request
+        .set_scheme(Some(&Scheme::Http))
+        .expect("setting scheme");
+    let addr = test_programs::p3::wasi::cli::environment::get_environment()
+        .into_iter()
+        .find_map(|(k, v)| k.eq("HTTP_SERVER").then_some(v))
+        .unwrap();
+    request
+        .set_authority(Some(&addr))
+        .expect("setting authority");
+    request
+        .set_path_with_query(Some("/"))
+        .expect("setting path with query");
+
+    (request, contents_tx, trailers_tx)
+}
+
+impl test_programs::p3::exports::wasi::cli::run::Guest for Component {
+    async fn run() -> Result<(), ()> {
+        {
+            println!("writing enough");
+            let (_, mut contents_tx, trailers_tx) = make_request();
+            contents_tx.send(b"long enough".to_vec()).await.unwrap();
+            drop(contents_tx);
+            trailers_tx.write(Ok(None)).await;
+        }
+
+        {
+            println!("writing too little");
+            let (_, mut contents_tx, trailers_tx) = make_request();
+            contents_tx.send(b"msg".to_vec()).await.unwrap();
+            drop(contents_tx);
+            trailers_tx.write(Ok(None)).await;
+
+            // TODO: Figure out how/if to represent this in wasip3
+            //let e = OutgoingBody::finish(outgoing_body, None)
+            //    .expect_err("finish should fail");
+
+            //assert!(
+            //    matches!(&e, ErrorCode::HttpRequestBodySize(Some(3))),
+            //    "unexpected error: {e:#?}"
+            //);
+        }
+
+        {
+            println!("writing too much");
+            let (_, mut contents_tx, trailers_tx) = make_request();
+            contents_tx
+                .send(b"more than 11 bytes".to_vec())
+                .await
+                .unwrap();
+            drop(contents_tx);
+            trailers_tx.write(Ok(None)).await;
+
+            // TODO: Figure out how/if to represent this in wasip3
```

Comment:
Not really sure what to do about this - can't see a way to directly translate this to wasip3. At least given current design it feels like assuming components themselves handle `content-length` is the right approach, but not sure. Either way, it feels like something to address in a follow-up

Reply:
As discussed with @pchickey offline, we'll just have to call `handle` here